src/main/java/com/stripe/util/CaseInsensitiveMap.java

```diff
@@ -0,0 +1,120 @@
+package com.stripe.util;
+
+import java.io.Serializable;
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A case-insensitive {@link Map}.
+ *
+ * <p>All keys are expected to be strings (though {@code null} is a valid key). The map remembers
+ * the case of the last key to be set, and {@link Map#keySet()} or {@link Map#entrySet()} will
```

Comment:
Just OOC, any particular reason to remember the case that was used? Just noting because the Ruby version that I wrote definitely does not do that :)

Reply:
I was influenced by Python's requests' implementation: https://github.com/psf/requests/blob/a9ee0eef5a70acb9ed35622b3675574b11f92cb4/requests/structures.py#L15. I agree it's probably unnecessary, but I thought it might be occasionally useful, for e.g. debugging.

Reply:
Fair enough! Seems like it could occasionally be useful, so +1.